output/cloud/expv2/collect.go

```diff
@@ -0,0 +1,158 @@
+	q.m.Lock()
+	defer q.m.Unlock()
+
+	if len(q.buckets) < 1 {
+		return nil
+	}
+	b := make([]timeBucket, len(q.buckets))
+	copy(b, q.buckets)
+	q.buckets = q.buckets[:0]
+	return b
+}
+
+// Insert adds an item in the queue.
+func (q *bucketQ) Push(b []timeBucket) {
+	q.m.Lock()
+	q.buckets = append(q.buckets, b...)
+	q.m.Unlock()
+}
+
+type timeBucket struct {
+	TimeSeries metrics.TimeSeries
+	Time       time.Time
+	Sink       metrics.Sink
+}
+
+type perSeriesBuckets map[metrics.TimeSeries]timeBucket
+
+func newPerSeriesBuckets() perSeriesBuckets {
+	return make(map[metrics.TimeSeries]timeBucket)
+}
+
+func (m perSeriesBuckets) AddSample(bucketID int64, s metrics.Sample, aggrPeriod int64) {
+	b, ok := m[s.TimeSeries]
+	if !ok {
+		b = timeBucket{
+			Time:       time.Unix(0, (bucketID*aggrPeriod)+(aggrPeriod/2)).Truncate(time.Microsecond),
+			TimeSeries: s.TimeSeries,
+			Sink:       newSink(s.Metric.Type),
+		}
+		m[s.TimeSeries] = b
+	}
+	b.Sink.Add(s)
+}
+
+type collector struct {
+	buf   *output.SampleBuffer
+	bq    bucketQ
+	nowfn func() time.Time
+
+	aggregationPeriod time.Duration
+	waitPeriod        time.Duration
+
+	// we should no longer have to handle metrics that have times long in the past. So instead of a
+	// map, we can probably use a simple slice (or even an array!) as a ring buffer to store the
+	// aggregation buckets. This should save us a some time, since it would make the lookups and WaitPeriod
+	// checks basically O(1). And even if for some reason there are occasional metrics with past times that
+	// don't fit in the chosen ring buffer size, we could just send them along to the buffer unaggregated
+	aggrBuckets map[int64]perSeriesBuckets
+}
+
+func newCollector(buf *output.SampleBuffer, aggrPeriod, waitPeriod time.Duration) (*collector, error) {
+	if aggrPeriod == 0 {
+		return nil, errors.New("aggregation period is not allowed to be zero")
+	}
+	return &collector{
+		buf:               buf,
+		bq:                bucketQ{},
+		nowfn:             time.Now,
+		aggrBuckets:       make(map[int64]perSeriesBuckets),
+		aggregationPeriod: aggrPeriod,
+		waitPeriod:        waitPeriod,
+	}, nil
+}
+
+// CollectSamples drain the buffer and collect all the samples.
+func (c *collector) CollectSamples() {
+	c.bucketing(c.buf.GetBufferedSamples())
+
+	expired := c.expiredBuckets()
+	if len(expired) > 0 {
+		c.bq.Push(expired)
+	}
+}
+
+func (c *collector) bucketID(t time.Time) int64 {
+	return t.UnixNano() / int64(c.aggregationPeriod)
+}
+
+func (c *collector) bucketing(containers []metrics.SampleContainer) {
+	// Distribute all newly buffered samples into related buckets
+	for _, sampleContainer := range containers {
+		samples := sampleContainer.GetSamples()
+
+		for i := 0; i < len(samples); i++ {
+			bucketID := c.bucketID(samples[i].Time)
+
+			// Get or create a time bucket
+			buckets, ok := c.aggrBuckets[bucketID]
+			if !ok {
+				buckets = newPerSeriesBuckets()
+				c.aggrBuckets[bucketID] = buckets
+			}
```

Comment:
future optimization that we can do here that I came up while writing [this comment](https://github.com/grafana/k6/pull/3071#discussion_r1197945597) 

It is very likely that `bucketID` will haver the same value for sequential samples so will `buckets` but will do map access for each of those.

So we can potentially save the last `buecketID` + `bueckts` values and check against the new ones first.

I would levea that for when we have the rest of the things figured out and when there are some good benchmarks to run against.